File: packages/modules/filesystem/android/src/main/java/run/stingjs/modules/filesystem/FilesystemModule.kt
package run.stingjs.modules.filesystem

import android.content.Context
import android.system.Os
import android.system.OsConstants
import java.io.File
import java.util.concurrent.Executors
import run.stingjs.runtime.StingNativeModule
import run.stingjs.runtime.StingNativeModuleCompletion
import run.stingjs.runtime.StingNativeModuleError
import run.stingjs.runtime.StingNativeModuleResult

class FilesystemModule(context: Context) : StingNativeModule {
    override val name = "Filesystem"
    override val version = "0.1.0"

    private val appContext = context.applicationContext
    private val executor = Executors.newCachedThreadPool { runnable ->
        Thread(runnable, "StingFilesystem").apply { isDaemon = true }
    }

    override fun callSync(method: String, arguments: List<Any?>): Any? {
        throw StingNativeModuleError(
            code = "E_SYNC_UNSUPPORTED",
            message = "Filesystem methods are asynchronous",
        )
    }

    override fun callAsync(
        method: String,
        arguments: List<Any?>,
        completion: StingNativeModuleCompletion,
    ) {
        executor.execute {
            try {
                completion(StingNativeModuleResult.Success(perform(method, arguments)))
            } catch (error: StingNativeModuleError) {
                completion(StingNativeModuleResult.Failure(error))
            } catch (error: Throwable) {
                completion(
                    StingNativeModuleResult.Failure(
                        StingNativeModuleError(
                            code = "E_IO",
                            message = error.message ?: "Filesystem operation failed",
                            details = mapOf("method" to method),
                        ),
                    ),
                )
            }
        }
    }

    private fun perform(method: String, arguments: List<Any?>): Any? = when (method) {
        "readText" -> {
            val target = target(arguments, pathIndex = 0, directoryIndex = 1)
            requireExistingFile(target)
            try {
                target.readText(Charsets.UTF_8)
            } catch (error: Throwable) {
                throw ioError("Unable to read file", target, error)
            }
        }

        "writeText" -> {
            val contents = arguments.getOrNull(1) as? String
                ?: throw invalidArgument("Filesystem.writeText requires string contents")
            val target = target(arguments, pathIndex = 0, directoryIndex = 2)
            requireWritableTarget(target)
            try {
                target.writeText(contents, Charsets.UTF_8)
                null
            } catch (error: Throwable) {
                throw ioError("Unable to write file", target, error)
            }
        }

        "delete" -> {
            val target = target(arguments, pathIndex = 0, directoryIndex = 1)
            if (!target.exists()) throw notFound(target)
            try {
                if (!deleteWithoutFollowingSymlinks(target)) {
                    throw StingNativeModuleError(
                        code = "E_IO",
                        message = "Unable to delete path",
                        details = mapOf("path" to target.path),
                    )
                }
                null
            } catch (error: StingNativeModuleError) {
                throw error
            } catch (error: Throwable) {
                throw ioError("Unable to delete path", target, error)
            }
        }

        "makeDirectory" -> {
            val target = target(arguments, pathIndex = 0, directoryIndex = 1)
            if (target.exists()) {
                throw StingNativeModuleError(
                    code = "E_ALREADY_EXISTS",
                    message = "Filesystem path already exists",
                    details = mapOf("path" to target.path),
                )
            }
            try {
                if (!target.mkdirs()) {
                    throw StingNativeModuleError(
                        code = "E_IO",
                        message = "Unable to create directory",
                        details = mapOf("path" to target.path),
                    )
                }
                null
            } catch (error: StingNativeModuleError) {
                throw error
            } catch (error: Throwable) {
                throw ioError("Unable to create directory", target, error)
            }
        }

        "getInfo" -> {
            val target = target(arguments, pathIndex = 0, directoryIndex = 1)
            if (!target.exists()) {
                mapOf(
                    "exists" to false,
                    "type" to null,
                    "size" to 0,
                    "modifiedAt" to null,
                )
            } else {
                mapOf(
                    "exists" to true,
                    "type" to if (target.isDirectory) "directory" else "file",
                    "size" to if (target.isDirectory) 0 else target.length(),
                    "modifiedAt" to target.lastModified().takeIf { it > 0L },
                )
            }
        }

        else -> throw StingNativeModuleError(
            code = "E_METHOD_NOT_FOUND",
            message = "Filesystem does not implement asynchronous method $method",
        )
    }

    private fun target(
        arguments: List<Any?>,
        pathIndex: Int,
        directoryIndex: Int,
    ): File {
        val path = arguments.getOrNull(pathIndex) as? String
            ?: throw invalidArgument("Filesystem requires a relative path")
        val directory = arguments.getOrNull(directoryIndex) as? String
            ?: throw invalidArgument("Filesystem requires a base directory")
        return resolve(path, directory)
    }

    private fun resolve(path: String, directory: String): File {
        if (path.isEmpty() || path.startsWith('/') || path.contains('\\')) {
            throw invalidPath(path)
        }
        val components = path.split('/')
        if (components.any { it.isEmpty() || it == "." || it == ".." }) {
            throw invalidPath(path)
        }

        val root = when (directory) {
            "documents" -> appContext.filesDir
            "cache" -> appContext.cacheDir
            else -> throw invalidArgument("Filesystem base directory must be documents or cache")
        }.canonicalFile

        val resolved = File(root, path).canonicalFile
        val rootPrefix = root.path.trimEnd(File.separatorChar) + File.separator
        if (!resolved.path.startsWith(rootPrefix)) {
            throw invalidPath(path)
        }
        return resolved
    }

    private fun requireExistingFile(target: File) {
        if (!target.exists()) throw notFound(target)
        if (target.isDirectory) {
            throw StingNativeModuleError(
                code = "E_IS_DIRECTORY",
                message = "Filesystem path is a directory",
                details = mapOf("path" to target.path),
            )
        }
    }

    private fun requireWritableTarget(target: File) {
        if (target.exists() && target.isDirectory) {
            throw StingNativeModuleError(
                code = "E_IS_DIRECTORY",
                message = "Filesystem path is a directory",
                details = mapOf("path" to target.path),
            )
        }
        val parent = target.parentFile
        if (parent == null || !parent.exists() || !parent.isDirectory) {
            throw StingNativeModuleError(
                code = "E_NOT_FOUND",
                message = "Filesystem parent directory does not exist",
                details = mapOf("path" to (parent?.path ?: target.path)),
            )
        }
    }

    private fun deleteWithoutFollowingSymlinks(target: File): Boolean {
        val mode = Os.lstat(target.path).st_mode
        if (OsConstants.S_ISLNK(mode) || !target.isDirectory) {
            return target.delete()
        }

        val children = target.listFiles() ?: return false
        for (child in children) {
            if (!deleteWithoutFollowingSymlinks(child)) return false
        }
        return target.delete()
    }

    private fun invalidArgument(message: String) = StingNativeModuleError(
        code = "E_INVALID_ARGUMENT",
        message = message,
    )

    private fun invalidPath(path: String) = StingNativeModuleError(
        code = "E_INVALID_PATH",
        message = "Filesystem paths must stay inside the selected app-private root",
        details = mapOf("path" to path),
    )

    private fun notFound(target: File) = StingNativeModuleError(
        code = "E_NOT_FOUND",
        message = "Filesystem path does not exist",
        details = mapOf("path" to target.path),
    )

    private fun ioError(message: String, target: File, underlying: Throwable) = StingNativeModuleError(
        code = "E_IO",
        message = "$message: ${underlying.message ?: underlying::class.java.simpleName}",
        details = mapOf("path" to target.path),
    )
}

File: packages/modules/filesystem/ios/FilesystemModule.swift
import Foundation
import StingRuntime

public final class FilesystemModule: StingNativeModule {
    public let name = "Filesystem"
    public let version = "0.1.0"

    private let fileManager: FileManager
    private let queue = DispatchQueue(
        label: "run.stingjs.modules.filesystem",
        qos: .utility,
        attributes: .concurrent
    )

    public init(fileManager: FileManager = .default) {
        self.fileManager = fileManager
    }

    public func callSync(method: String, arguments: [Any]) throws -> Any? {
        throw StingNativeModuleError(
            code: "E_SYNC_UNSUPPORTED",
            message: "Filesystem methods are asynchronous"
        )
    }

    public func callAsync(
        method: String,
        arguments: [Any],
        completion: @escaping StingNativeModuleCompletion
    ) {
        queue.async { [self] in
            do {
                completion(.success(try perform(method: method, arguments: arguments)))
            } catch let error as StingNativeModuleError {
                completion(.failure(error))
            } catch {
                completion(.failure(StingNativeModuleError(
                    code: "E_IO",
                    message: error.localizedDescription,
                    details: ["method": method]
                )))
            }
        }
    }

    private func perform(method: String, arguments: [Any]) throws -> Any? {
        switch method {
        case "readText":
            let target = try target(arguments: arguments, pathIndex: 0, directoryIndex: 1)
            try requireExistingFile(target)
            do {
                return try String(contentsOf: target, encoding: .utf8)
            } catch {
                throw ioError("Unable to read file", path: target.path, underlying: error)
            }

        case "writeText":
            guard arguments.count > 1, let contents = arguments[1] as? String else {
                throw invalidArgument("Filesystem.writeText requires string contents")
            }
            let target = try target(arguments: arguments, pathIndex: 0, directoryIndex: 2)
            try requireWritableTarget(target)
            do {
                try Data(contents.utf8).write(to: target, options: .atomic)
                return nil
            } catch {
                throw ioError("Unable to write file", path: target.path, underlying: error)
            }

        case "delete":
            let target = try target(arguments: arguments, pathIndex: 0, directoryIndex: 1)
            guard fileManager.fileExists(atPath: target.path) else {
                throw notFound(target)
            }
            do {
                try fileManager.removeItem(at: target)
                return nil
            } catch {
                throw ioError("Unable to delete path", path: target.path, underlying: error)
            }

        case "makeDirectory":
            let target = try target(arguments: arguments, pathIndex: 0, directoryIndex: 1)
            guard !fileManager.fileExists(atPath: target.path) else {
                throw StingNativeModuleError(
                    code: "E_ALREADY_EXISTS",
                    message: "Filesystem path already exists",
                    details: ["path": target.path]
                )
            }
            do {
                try fileManager.createDirectory(
                    at: target,
                    withIntermediateDirectories: true
                )
                return nil
            } catch {
                throw ioError("Unable to create directory", path: target.path, underlying: error)
            }

        case "getInfo":
            let target = try target(arguments: arguments, pathIndex: 0, directoryIndex: 1)
            var isDirectory: ObjCBool = false
            guard fileManager.fileExists(atPath: target.path, isDirectory: &isDirectory) else {
                return [
                    "exists": false,
                    "type": NSNull(),
                    "size": 0,
                    "modifiedAt": NSNull(),
                ] as [String: Any]
            }

            do {
                let attributes = try fileManager.attributesOfItem(atPath: target.path)
                let size = isDirectory.boolValue
                    ? 0
                    : (attributes[.size] as? NSNumber)?.doubleValue ?? 0
                let modifiedAt: Any
                if let date = attributes[.modificationDate] as? Date {
                    modifiedAt = date.timeIntervalSince1970 * 1_000
                } else {
                    modifiedAt = NSNull()
                }
                return [
                    "exists": true,
                    "type": isDirectory.boolValue ? "directory" : "file",
                    "size": size,
                    "modifiedAt": modifiedAt,
                ] as [String: Any]
            } catch {
                throw ioError("Unable to inspect path", path: target.path, underlying: error)
            }

        default:
            throw StingNativeModuleError(
                code: "E_METHOD_NOT_FOUND",
                message: "Filesystem does not implement asynchronous method \(method)"
            )
        }
    }

    private func target(
        arguments: [Any],
        pathIndex: Int,
        directoryIndex: Int
    ) throws -> URL {
        guard arguments.count > pathIndex, let path = arguments[pathIndex] as? String else {
            throw invalidArgument("Filesystem requires a relative path")
        }
        guard arguments.count > directoryIndex, let directory = arguments[directoryIndex] as? String else {
            throw invalidArgument("Filesystem requires a base directory")
        }
        return try resolve(path: path, directory: directory)
    }

    private func resolve(path: String, directory: String) throws -> URL {
        guard !path.isEmpty,
              !path.hasPrefix("/"),
              !path.contains("\\") else {
            throw invalidPath(path)
        }

        let components = path.split(separator: "/", omittingEmptySubsequences: false)
        guard !components.contains(where: { $0.isEmpty || $0 == "." || $0 == ".." }) else {
            throw invalidPath(path)
        }

        let root: URL
        switch directory {
        case "documents":
            guard let value = fileManager.urls(for: .documentDirectory, in: .userDomainMask).first else {
                throw StingNativeModuleError(
                    code: "E_IO",
                    message: "Application documents directory is unavailable"
                )
            }
            root = value
        case "cache":
            guard let value = fileManager.urls(for: .cachesDirectory, in: .userDomainMask).first else {
                throw StingNativeModuleError(
                    code: "E_IO",
                    message: "Application cache directory is unavailable"
                )
            }
            root = value
        default:
            throw invalidArgument("Filesystem base directory must be documents or cache")
        }

        let canonicalRoot = root.standardizedFileURL.resolvingSymlinksInPath()
        let canonicalTarget = canonicalRoot
            .appendingPathComponent(path)
            .standardizedFileURL
            .resolvingSymlinksInPath()
        let rootPrefix = canonicalRoot.path.hasSuffix("/")
            ? canonicalRoot.path
            : canonicalRoot.path + "/"

        guard canonicalTarget.path.hasPrefix(rootPrefix) else {
            throw invalidPath(path)
        }
        return canonicalTarget
    }

    private func requireExistingFile(_ target: URL) throws {
        var isDirectory: ObjCBool = false
        guard fileManager.fileExists(atPath: target.path, isDirectory: &isDirectory) else {
            throw notFound(target)
        }
        guard !isDirectory.boolValue else {
            throw StingNativeModuleError(
                code: "E_IS_DIRECTORY",
                message: "Filesystem path is a directory",
                details: ["path": target.path]
            )
        }
    }

    private func requireWritableTarget(_ target: URL) throws {
        var isDirectory: ObjCBool = false
        if fileManager.fileExists(atPath: target.path, isDirectory: &isDirectory), isDirectory.boolValue {
            throw StingNativeModuleError(
                code: "E_IS_DIRECTORY",
                message: "Filesystem path is a directory",
                details: ["path": target.path]
            )
        }

        let parent = target.deletingLastPathComponent()
        var parentIsDirectory: ObjCBool = false
        guard fileManager.fileExists(atPath: parent.path, isDirectory: &parentIsDirectory),
              parentIsDirectory.boolValue else {
            throw StingNativeModuleError(
                code: "E_NOT_FOUND",
                message: "Filesystem parent directory does not exist",
                details: ["path": parent.path]
            )
        }
    }

    private func invalidArgument(_ message: String) -> StingNativeModuleError {
        StingNativeModuleError(code: "E_INVALID_ARGUMENT", message: message)
    }

    private func invalidPath(_ path: String) -> StingNativeModuleError {
        StingNativeModuleError(
            code: "E_INVALID_PATH",
            message: "Filesystem paths must stay inside the selected app-private root",
            details: ["path": path]
        )
    }

    private func notFound(_ target: URL) -> StingNativeModuleError {
        StingNativeModuleError(
            code: "E_NOT_FOUND",
            message: "Filesystem path does not exist",
            details: ["path": target.path]
        )
    }

    private func ioError(_ message: String, path: String, underlying: Error) -> StingNativeModuleError {
        StingNativeModuleError(
            code: "E_IO",
            message: "\(message): \(underlying.localizedDescription)",
            details: ["path": path]
        )
    }
}
